Mathlib/GroupTheory/GroupAction/DomAct.lean

```diff
@@ -0,0 +1,225 @@
+
+## Implementation details
+
+### Motivation
+
+Right action can be represented in `mathlib` as an action of the opposite group `Mᵐᵒᵖ`. However,
+this "domain shift" action cannot be an instance because this would create a "diamond"
+(a.k.a. ambiguous notation): if `M` is a monoid, then how does `Mᵐᵒᵖ` act on `M → M`? On the one
+hand, `Mᵐᵒᵖ` acts on `M` by `c • a = a * c.unop`, thus we have an action
+`(c • f) a = f a * c.unop`. On the other hand, `M` acts on itself by multiplication on the left, so
+with this new instance we would have `(c • f) a = f (c.unop * a)`. Clearly, these are two different
+actions, so both of them cannot be instances in the library.
+
+To overcome this difficulty, we introduce a type synonym `DomMulAct M := Mᵐᵒᵖ` (notation:
+`Mᵈᵐᵃ`). This new type carries the same algebraic structures as `Mᵐᵒᵖ` but acts on `α → β` by this
+new action. So, e.g., `Mᵈᵐᵃ` acts on `(M → M) → M` by `DomMulAct.mk c • F f = F (fun a ↦ c • f a)`
+while `(Mᵈᵐᵃ)ᵈᵐᵃ` (which is isomorphic to `M`) acts on `(M → M) → M` by
+`DomMulAct.mk (DomMulAct.mk c) • F f = F (fun a ↦ f (c • a))`.
+
+### Action on bundled homomorphisms
+
+If the action of `M` on `A` preserves some structure, then `Mᵈᵐᵃ` acts on bundled homomorphisms from
+`A` to any type `B` that preserve the same structure. Examples (some of them are not yet in the
+library) include:
+
+- a `MulDistribMulAction` generates an action on `A →* B`;
+- a `DistribMulAction` generates an action on `A →+ B`;
+- an action on `α` that commutes with action of some other monoid `N` generates an action on
+  `α →[N] β`;
+- a `DistribMulAction` on an `R`-module that commutes with scalar multiplications by `c : R`
+  generates an action on `R`-linear maps from this module;
+- a continuous action on `X` generates an action on `C(X, Y)`;
+- a measurable action on `X` generates an action on `{ f : X → Y // Measurable f }`;
+- a quasi measure preserving action on `X` generates an action on `X →ₘ[μ] Y`;
+- a measure preserving action generates an isometric action on `MeasureTheory.Lp _ _ _`.
+
+## Keywords
+
+group action, function, domain
+-/
+open Function
+
+/-- If `M` multiplicatively acts on `α`, then `DomMulAct M` acts on `α → β` as well as some
+bundled maps from `α`. This is a type synonym for `MulOpposite M`, so this corresponds to a right
+action of `M`. -/
+@[to_additive "If `M` additively acts on `α`, then `DomAddAct M` acts on `α → β` as
+well as some bundled maps from `α`. This is a type synonym for `AddOpposite M`, so this corresponds
+to a right action of `M`."]
+def DomMulAct (M : Type _) := MulOpposite M
+
+@[inherit_doc] postfix:max "ᵈᵐᵃ" => DomMulAct
+@[inherit_doc] postfix:max "ᵈᵃᵃ" => DomAddAct
+
+namespace DomMulAct
+
+/-- Equivalence between `M` and `Mᵈᵐᵃ`. -/
+@[to_additive "Equivalence between `M` and `Mᵈᵐᵃ`."]
+def mk : M ≃ Mᵈᵐᵃ := MulOpposite.opEquiv
+
+/-!
+### Copy instances from `Mᵐᵒᵖ`
+-/
+
+set_option hygiene false in
```

Comment:
I'm curious if there's a hygienic way to do this.

Reply:
If this is the only reason not to merge this PR, then I can replace this macro with N lines it unfolds to.

Reply:
I think this is a nice trick; I'm just curious about what makes it unhygienic.

Reply:
Should we merge it then?